presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt

```diff
@@ -0,0 +1,437 @@
+
+        fun getService(): TextToSpeechService {
+            return this@TextToSpeechService
+        }
+    }
+
+    private data class FutureSpeechData(val speechMap: MutableMap<Long, Pair<String, Boolean>>,
+            val duration: Duration,
+            val countdown: LiveData<Long>) {
+
+        fun hasQueuedSpeech(): Boolean {
+            // if speechMap != null duration != null and countdown != null
+            val currentOffset = duration.seconds - (countdown.value ?: 0L)
+            return speechMap.keys.any { offset -> offset > currentOffset }
+        }
+    }
+
+    companion object {
+        private val LOGGER: Logger = LoggerFactory.getLogger(TextToSpeechService::class.java)
+    }
+
+    private val serviceBinder: Binder = Binder()
+    // values set when after a step is set by the client
+    private var textToSpeech: TextToSpeech? = null
+    private var futureSpeechData: FutureSpeechData? = null
+    private var textToSpeakOnInit: String? = null
+    // TODO rkolmos 10/19/2018 make this a dagger binding so users can add their own keys.
+    private val specialKeyMap: MutableMap<String, (Duration) -> Long> = mutableMapOf()
+
+    init {
+        specialKeyMap["start"] = { _ -> 0 }
+        // halfway rounding down to the nearest second is good enough.
+        specialKeyMap["halfway"] = { duration -> duration.seconds / 2 }
+        specialKeyMap["end"] = { duration -> duration.seconds }
+    }
+
+    // Queuing behavior is passed to all calls to TextToSpeech.speak(), default value is QUEUE_ADD
+    var queueingBehavior = TextToSpeech.QUEUE_ADD
+    // The duration of the sound played by playSound() and the vibration triggered by vibrate() in ms, default value
+    // is 500.
+    var defaultVibrateAndSoundDurationMillis: Long = 500
+    // The amount of time in seconds that speeches will still be spoken for if the service is initialized after
+    // the speeches exact time. (ie, if speechGracePeriod = 5 and the service is initialized 4 seconds after the step
+    // formally starts, instructions with the key "start" will still be spoken.
+    var speechGracePeriod: Long = 5
+    // The public getter for state should have type live data so client's don't modify it.
+    val state: LiveData<TextToSpeechState>
+        get() = _state
+    // The private state value is mutable so the service can update it's state as needed
+    private val _state = MutableLiveData<TextToSpeechState>()
+    private val bindCount = MutableLiveData<Int>()
+    private val cleanupMediator = MediatorLiveData<Pair<Int, TextToSpeechState>>()
+    private val cleanupObserver: Observer<Pair<Int, TextToSpeechState>>
+
+    init {
+        bindCount.postValue(0)
+        _state.postValue(TextToSpeechState(IDLE, null))
+        // We observe the state so that if every client unbinds and the service stops speaking, the service
+        // stops itself.
+        cleanupMediator.postValue(Pair(0, TextToSpeechState(IDLE, null)))
```

Comment:
It's hard to follow this because it doesn't seem explained anywhere what the first val of the Pair is.